man/crunchMake.1
.\" Automatically generated by Pandoc 2.9.2.1
.\"
.TH "CRUNCHMAKE" "1" "" "crunch 1.0.0" "crunch unit testing framework"
.hy
.SH NAME
.PP
\f[B]crunchMake\f[R] - crunch and crunch++ test builder
.SH SYNOPSIS
.PP
\f[B]crunchMake\f[R] [\f[B]-h\f[R]|\f[B]--help\f[R]]
.PD 0
.P
.PD
\f[B]crunchMake\f[R] [\f[B]-v\f[R]|\f[B]--version\f[R]]
.PD 0
.P
.PD
\f[B]crunchMake\f[R] [\f[I]options\f[R]] \f[I]FILES\f[R]
.SH DESCRIPTION
.PP
crunchMake is a utility to build crunch and crunch++ unit tests.
.SH OPTIONS
.TP
-v, --version
Prints the version information for crunch
.TP
-h, --help
Prints this help message
.SS Compiler support
.TP
-l\f[B]library\f[R]
Adds \f[B]library\f[R] to the compiler library linking set
.TP
-I\f[B]dir\f[R]
Adds \f[B]dir\f[R] to the compiler include search path
.TP
-D\f[B]macro\f[R]
Adds \f[B]macro\f[R] to the compiler\[cq]s macro predefinitions
.TP
-L\f[B]dir\f[R]
Adds \f[B]dir\f[R] to the compiler library search path
.TP
-o \f[B]file\f[R]
Use \f[B]file\f[R] for the output test library
.TP
-pthread
Specify that you wish to build and link against pthreads
.TP
-Wl,\f[B]option\f[R]
Adds \f[B]option\f[R] to the compiler-handled linker options
.TP
-std=\f[B]standard\f[R]
Sets the C or C++ standard to \f[B]standard\f[R].
In C++ mode, the minimimum supported standard is C++11.
In C mode, the minimum supported standard is C89.
.RS
.PP
This option must be specified in the form of either c\f[B]NN\f[R] or
c++\f[B]NN\f[R] where \f[B]NN\f[R] is the version number to use.
.RE
.TP
-z \f[B]keyword\f[R]
Specifies direct linker options - the exact meaning and options
available depend on your compiler\[cq]s specific linker.
.TP
--coverage
Enables linking against code that has been code-coverage enabled
.TP
--debug
Specifies that you wish the test to be built in debug mode.
crunchMake defaults to building tests with level 2 (or equivilent)
optimisations.
.TP
-fsanitize=\f[B]sanitizers\f[R]
Enable the comma separated list of sanitizers \f[B]sanitizers\f[R] on
the test being built
.TP
-flto
Enables Link Time Optimisation on the test\[cq]s build
.RS
.PP
If you have LTO enabled on your build and are passing in LTO-compiled
objects, you MUST provide this flag.
.RE
.SS Utility output options
.TP
--log
Tells the engine to log all test output to the file named
.TP
-s, --silent
Silences all output from crunchMake, leaving only compiler output on
stdout and stderr
.TP
-q, --quiet
Reduces output from crunchMake to kernel Makefile like output such as
\[cq] CCLD test.cxx => test.so\[cq]
.RS
.PP
By default, the utility will verbosely dump the full compiler
invocations it is running, sticking to conventions used by other
\f[C]make\f[R]-like utilities.
.RE
.PP
The files given on the crunchMake command line are gathered together and
passed as inputs to the compiler.
.PD 0
.P
.PD
This allows you to specify multiple TUs and additional object files to
be compiled and linked to the suite to produce a complete library
.SH FURTHER INFORMATION
.PP
\f[B]crunchMake\f[R] is a tool that aims to ensure a working build of
your tests without having to worry about exactly where
\f[B]\f[CB]crunch\f[B](1)\f[R] or \f[B]\f[CB]crunch++\f[B](1)\f[R] is
installed or how it was built.
.PP
It provides transparency for many compiler options, and platform- and
compiler-specific translations for the rest.
.PP
The important translated options when building test suites are:
.TP
--coverage
This option enables the compiler-specific code coverage options for the
build for when doing a code-coverage enabled build of your project
.TP
--debug
This option enables debugging information on the test suite to allow
setting breakpoints in the tests and inspecting state.
.RS
.PP
Example usage of such a build:
\f[B]\f[CB]gdb --args crunch++ testSuite\f[B]\f[R]
.RE
.PP
When compiling C++ test suites, \f[B]crunchMake\f[R] will automatically
feed the compiler with the visibility options
\f[B]\f[CB]-fvisbility-inlines-hidden\f[B]\f[R] and
\f[B]\f[CB]-fvisibility=hidden\f[B]\f[R] on GCC-like compilers.
.SS A simple crunch++ test suite
.PP
The following example can be used as a template for a minimum viable
useful test suite written using crunch++:
.IP
.nf
\f[C]
#include <crunch++.h>

class testSuite final : public testsuite
{
private:
    void testCase() { }

public:
    void registerTests() final
    {
        CRUNCHpp_TEST(testCase)
    }
};

CRUNCHpp_TESTS(testSuite)
\f[R]
.fi
.PP
Assuming this has been written out as test.cxx, compilation and
excecution of this example looks like:
.IP
.nf
\f[C]
$ crunchMake test.cxx
c++ -fPIC -DPIC -fvisibility=hidden -fvisibility-inlines-hidden -std=c++11  test.cxx -shared -I/usr/include -L/usr/lib -Wl,-rpath,/usr/lib -lcrunch++ -O2 -pthread -o test.so

$ crunch++ test
Running test suite test...
Running tests in class 9testSuite...
testCase...                                                                          [  OK  ]
Total tests: 1,  Failures: 0,  Pass rate: 100.00%
\f[R]
.fi
.SS A simple crunch test suite
.PP
The following example can be used as a template for a minimum viable
useful test suite written using crunch:
.IP
.nf
\f[C]
#include <crunch.h>

void testCase() { }

BEGIN_REGISTER_TESTS()
    TEST(testCase)
END_REGISTER_TESTS()
\f[R]
.fi
.PP
Assuming this has been written out as test.c, compilation and excecution
of this example looks like:
.IP
.nf
\f[C]
$ crunchMake test.c
cc -fPIC -DPIC  test.c -shared -I/usr/include -L/usr/lib -Wl,-rpath,/usr/lib -lcrunch -O2 -pthread -o test.so

$ crunch test
Running test suite test...
testCase...                                                                          [  OK  ]
Total tests: 1,  Failures: 0,  Pass rate: 100.00%
\f[R]
.fi
.SH BUGS
.PP
Report bugs using <https://github.com/DX-MON/crunch/issues>
.SH AUTHORS
.PP
Rachel Mant <dx-mon@users.sourceforge.net>
.SH SEE ALSO
.PP
\f[B]\f[CB]crunch\f[B](1)\f[R], \f[B]\f[CB]crunch++\f[B](1)\f[R]


crunchMake/crunchMake.cpp
// SPDX-License-Identifier: LGPL-3.0-or-later
#include <cstring>
#include <cstdlib>
#ifndef _MSC_VER
#include <unistd.h>
#else
#include <io.h>
#define R_OK 6
#endif
#include <array>
#include <vector>
#include <string>
#include <memory>
#include <algorithm>
#include <numeric>
#include <substrate/utility>
#include <core.hxx>
#include <logger.hxx>
#include <argsParser.hxx>
#include <stringFuncs.hxx>
#include "crunchCompiler.hxx"
#include <crunchMake.h>
#include <version.hxx>

namespace crunch
{
	using internal::stringView;

	parsedArgs_t parsedArgs;
	std::vector<internal::stringView> inclDirs, libDirs;
	std::vector<internal::stringView> linkLibs, linkObjs;
	std::vector<std::string> linkArgs;
	std::vector<internal::stringView> tests;
	uint32_t numLinkArgs = 0;

	constexpr static auto exts{substrate::make_array<stringView>({
		".c"_sv, ".cpp"_sv, ".cc"_sv, ".cxx"_sv, ".i"_sv, ".s"_sv, ".S"_sv, ".sx"_sv, ".asm"_sv
	})};
	constexpr static auto cxxExts{substrate::make_array<stringView>({".cpp"_sv, ".cc"_sv, ".cxx"_sv})};
	constexpr static auto objExts{substrate::make_array<stringView>({".o"_sv, ".obj"_sv, ".a"_sv})};

	std::string inclDirFlags{}, libDirFlags{}, objs{}, libs{};
	bool silent, quiet, pthread, codeCoverage, debugBuild;

	constexpr static auto args{substrate::make_array<arg_t>(
	{
		{"--help"_sv, 0, 0, 0},
		{"-h"_sv, 0, 0, 0},
		{"--version"_sv, 0, 0, 0},
		{"-v"_sv, 0, 0, 0},
		{"-l"_sv, 0, 0, ARG_REPEATABLE | ARG_INCOMPLETE},
		{"-I"_sv, 0, 0, ARG_REPEATABLE | ARG_INCOMPLETE},
		{"-D"_sv, 0, 0, ARG_REPEATABLE | ARG_INCOMPLETE},
		{"-L"_sv, 0, 0, ARG_REPEATABLE | ARG_INCOMPLETE},
		{"-o"_sv, 1, 1, 0},
		{"--log"_sv, 1, 1, 0},
		{"--silent"_sv, 0, 0, 0},
		{"-s"_sv, 0, 0, 0},
		{"--quiet"_sv, 0, 0, 0},
		{"-q"_sv, 0, 0, 0},
		{"-pthread"_sv, 0, 0, 0},
		{"-Wl,"_sv, 0, 0, ARG_INCOMPLETE},
		{"-std="_sv, 0, 0, ARG_INCOMPLETE},
		{"-z"_sv, 1, 1, ARG_REPEATABLE},
		{"--coverage"_sv, 0, 0, 0},
		{"--debug"_sv, 0, 0, 0},
		{"-fsanitize="_sv, 0, 0, ARG_INCOMPLETE},
		{"-flto"_sv, 0, 0, 0},
		{{}, 0, 0, 0}
	})};

	template<size_t N> constexpr bool checkExt(const internal::stringView &file, const std::array<stringView, N> &exts) noexcept
	{
		constexpr auto npos{internal::stringView::npos};
		const auto dot{file.rfind('.')};
		if (dot == npos)
			return false;
		return std::find_if(exts.begin(), exts.end(),
			[=](const stringView ext) { return file.compare(dot, npos, ext) == 0; }
		) != exts.end();
	}

	constexpr bool validExt(const internal::stringView &file)
		{ return checkExt(file, exts); }
	bool isCXX(const internal::stringView &file)
		{ return checkExt(file, cxxExts); }
	constexpr bool isObj(const internal::stringView &file)
		{ return checkExt(file, objExts); }

	bool getTests()
	{
		for (const auto &parsedArg : parsedArgs)
		{
			const auto &value = parsedArg.value;
			if (!findArgInArgs(value) && !isObj(value))
				tests.emplace_back(value);
		}
		return !tests.empty();
	}

	inline void getLinkFunc(std::vector<internal::stringView> &var, const char *find)
	{
		for (const auto &parsedArg : parsedArgs)
		{
			const auto &value = parsedArg.value;
			if (strncmp(value.data(), find, 2) == 0)
				var.emplace_back(value);
		}
	}

	void getInclDirs()
	{
		getLinkFunc(inclDirs, "-I");
		getLinkFunc(inclDirs, "-D");
	}
	void getLinkLibs() { getLinkFunc(linkLibs, "-l"); }
	void getLibDirs() { getLinkFunc(libDirs, "-L"); }

	void getLinkObjs()
	{
		for (const auto &parsedArg : parsedArgs)
		{
			const auto &value = parsedArg.value;
			if (!findArgInArgs(value) && isObj(value))
				linkObjs.emplace_back(value);
		}
	}

	inline std::string argToString(const parsedArg_t &arg)
	{
		auto result{arg.value.toString() + ' '};
		for (uint32_t i = 0; i < arg.paramsFound; ++i)
			result += arg.params[i] + ' ';
		return result;
	}

	void getLinkArgs()
	{
		for (const auto &parsedArg : parsedArgs)
		{
			if (parsedArg.matches("-z"_sv) || parsedArg.matches("-Wl,"_sv) ||
				parsedArg.matches("-flto"_sv))
				linkArgs.emplace_back(argToString(parsedArg));
		}
	}

	std::string toO(const std::string &file)
	{
		const auto dotPos{file.find_last_of('.')};
		auto objFile{file.substr(0, dotPos)};
#ifndef _MSC_VER
		return objFile += ".o"s;
#else
		return objFile += ".obj"s;
#endif
	}

	std::string computeObjName(const std::string &file)
	{
		const auto *const output{findArg(parsedArgs, "-o"_sv, nullptr)};
		if (output)
			return toO(output->params[0]);
		return toO(file);
	}

	std::string toSO(const std::string &file)
	{
		const auto dotPos{file.find_last_of('.')};
		auto soFile{file.substr(0, dotPos)};
		return soFile += libExt;
	}

	std::string computeSOName(const std::string &file)
	{
		const auto *const output{findArg(parsedArgs, "-o"_sv, nullptr)};
		if (output)
			return output->params[0];
		return toSO(file);
	}

	inline std::string argsToString(const std::vector<internal::stringView> &var)
	{
		return std::accumulate(var.begin(), var.end(), std::string{},
			[](const std::string &result, const internal::stringView &value)
				{ return result + value.toString() + ' '; }
		);
	}

	inline std::string argsToString(const std::vector<std::string> &var)
	{
		return std::accumulate(var.begin(), var.end(), std::string{},
			[](const std::string &result, const std::string &value)
				{ return result + value + ' '; }
		);
	}

#ifdef _MSC_VER
	string convertLibDirs()
	{
		std::string ret{};
		for (const auto &value : libDirs)
			ret += "-libpath:" + value.substr(2).toString() + ' ';
		return ret;
	}
#endif

	void inclDirFlagsToString() { inclDirFlags = argsToString(inclDirs); }
#ifndef _MSC_VER
	void libDirFlagsToString() { libDirFlags = argsToString(libDirs); }
#else
	void libDirFlagsToString() { libDirFlags = convertLibDirs(); }
#endif
	void objsToString() { objs = argsToString(linkObjs); }
	void libsToString() { libs = argsToString(linkLibs) + argsToString(linkArgs); }

#ifndef _MSC_VER
	internal::stringView standardVersion(constParsedArg_t version)
	{
		if (!version)
			return "-std=c++11"_sv;
		const auto *const str = version->value.data() + 5;
		if (strlen(str) != 5 || strncmp(str, "c++", 3) != 0 || str[3] == '8' || str[3] == '9')
		{
			testPrintf("Warning, standard version must be at least C++11");
			return "-std=c++11"_sv;
		}
		return version->value;
	}

	void buildCXXString()
	{
		const auto *const standard{findArg(parsedArgs, "-std="_sv, nullptr)};
		cxxCompiler += standardVersion(standard).toString() + ' ';
	}
#endif

	void handleSanitizers()
	{
		const auto *const sanitizer{findArg(parsedArgs, "-fsanitize="_sv, nullptr)};
		if (!sanitizer)
			return;
		const stringView sanitizers{sanitizer->value.data() + 11, sanitizer->value.length() - 11};
		for (size_t offset{}; offset < sanitizers.length();)
		{
			const auto length{[](const stringView &value, const size_t offset) noexcept
			{
				const auto result{value.find(',', offset)};
				return (result == stringView::npos ? value.length() : result) - offset;
			}(sanitizers, offset)};
			const auto option{"-fsanitize="_s + sanitizers.substr(offset, length).toString() + ' '};
			cCompiler += option;
			cxxCompiler += option;
			offset += length + 1;
		}
	}

	int compileTests()
	{
		int32_t ret = 0;
		inclDirFlagsToString();
		libDirFlagsToString();
		objsToString();
		libsToString();
#ifndef _MSC_VER
		buildCXXString();
#endif
		testLog *logFile = nullptr;
		const auto *const logParam{findArg(parsedArgs, "--log"_sv, nullptr)};
		const auto logging = bool(logParam);
		if (logging)
			logFile = startLogging(logParam->params[0].data());
		if (!silent)
			silent = bool(findArg(parsedArgs, "-s"_sv, nullptr));
		if (!quiet)
			quiet = bool(findArg(parsedArgs, "-q"_sv, nullptr));

		handleSanitizers();

		for (const auto &test : tests)
		{
			if (access(test.data(), R_OK) == 0 && validExt(test))
			{
				ret = compileTest(test.toString());
				if (ret)
					break;
			}
			else
				testPrintf("Error, %s does not exist, skipping..\n", test.data());
		}
		if (logging)
			stopLogging(logFile);
		return ret;
	}

	bool handleVersionOrHelp()
	{
		constParsedArg_t version{findArg(parsedArgs, "--version"_sv, nullptr)};
		constParsedArg_t versionShort{findArg(parsedArgs, "-v"_sv, nullptr)};
		constParsedArg_t help{findArg(parsedArgs, "--help"_sv, nullptr)};
		constParsedArg_t helpShort{findArg(parsedArgs, "-h"_sv, nullptr)};

		if (help || helpShort)
			puts(crunchpp::help.data());
		else if (version || versionShort)
			testPrintf("crunchMake %s (%s %s %s-%s)\n", crunchpp::version.data(), crunchpp::compiler.data(),
				crunchpp::compilerVersion.data(), crunchpp::system.data(), crunchpp::arch.data());
		else
			return false;
		return true;
	}

	int32_t main(const int32_t argc, const char *const *const argv) try
	{
		registerArgs(args.data());
		parsedArgs = parseArguments(argc, argv);
		if (!parsedArgs.empty() && handleVersionOrHelp())
			return 0;
		else if (parsedArgs.empty() || !getTests())
		{
			testPrintf("Fatal error: There are no source files to build given on the command line!\n");
			return 2;
		}
		getInclDirs();
		getLibDirs();
		getLinkLibs();
		getLinkObjs();
		getLinkArgs();
		silent = bool(findArg(parsedArgs, "--silent"_sv, nullptr));
		quiet = bool(findArg(parsedArgs, "--quiet"_sv, nullptr));
		pthread = bool(findArg(parsedArgs, "-pthread"_sv, nullptr));
		codeCoverage = bool(findArg(parsedArgs, "--coverage"_sv, nullptr));
		debugBuild = bool(findArg(parsedArgs, "--debug"_sv, nullptr));
		return compileTests();
	}
	catch (const std::out_of_range &error)
	{
		testPrintf("Error during string splicing operations: %s\n", error.what());
		return 2;
	}
} // namespace crunch

int main(int argc, char **argv) { return crunch::main(argc, argv); }
